### PR TITLE
Assorted fixes

### PR DESCRIPTION
**Pull request by Ralith**
Still some work to do, but this is a large step in the right direction. No issues found in Quinn itself yet.

### CODE REVIEW

File: quinn-interop/src/client.rs [hunk 1]
```diff
@@ -225,7 +225,7 @@ async fn h3_download_all(
 fn config(alpn: &str, suites: CipherSuite) -> Result<ClientConfig, Box<dyn std::error::Error>> {
     let mut provider = rustls::crypto::ring::default_provider();
     if let CipherSuite::Chacha20 = suites {
-        provider.cipher_suites = vec![TLS13_CHACHA20_POLY1305_SHA256];
+        provider.cipher_suites = vec![TLS13_CHACHA20_POLY1305_SHA256, TLS13_AES_128_GCM_SHA256];
```
Comment (djc) [inline quinn-interop/src/client.rs]: Nit: maybe add a comment why we need AES-128-GCM here?

File: quinn-interop/src/client.rs [hunk 2]
```diff
@@ -89,42 +89,50 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         std::process::exit(127);
     };
 
-    match std::env::var("TESTCASE").as_ref().map(String::as_str) {
-        Ok("http3") => {
+    let testcase = match std::env::var("TESTCASE") {
```
Comment (djc) [inline quinn-interop/src/client.rs]: Nit: `testcase` -> `test_case`?

File: quinn-interop/src/client.rs [hunk 3]
```diff
@@ -168,21 +170,25 @@ async fn hq_download_all(
     conn: quinn::Connection,
     requests: &[http::Uri],
 ) -> Result<(), Box<dyn std::error::Error>> {
-    // TODO spawn
-    future::try_join_all(requests.into_iter().cloned().map(move |req| {
+    for req in requests.into_iter().cloned() {
```
Comment (djc) [inline quinn-interop/src/client.rs]: Nit: this commit has kind of a lot going on, maybe make this three commits?

- Extract `hq_download()`
- Extract `h3_download()`
- Get rid of futures dependency

File: quinn-interop/src/client.rs [hunk 4]
```diff
@@ -264,7 +264,11 @@ fn config(alpn: &str, suites: CipherSuite) -> Result<ClientConfig, Box<dyn std::
         .max_idle_timeout(Some(Duration::from_millis(9000).try_into()?))
         .initial_rtt(Duration::from_millis(100))
         // https://github.com/quic-interop/quic-interop-runner/issues/397
-        .enable_segmentation_offload(false);
+        .enable_segmentation_offload(false)
+        // Don't bother probing a known network environment, and avoid
+        // https://github.com/quic-interop/quic-interop-runner/issues/398
+        .mtu_discovery_config(None)
+        .initial_mtu(1500 - 40 - 8);
```
Comment (djc) [inline quinn-interop/src/client.rs]: Add a comment on where `1500 - 40 - 8` comes from?

File: quinn-interop/src/client.rs [hunk 5]
```diff
@@ -18,6 +18,29 @@ use rustls::{
 use tokio::{self, fs::File, io::AsyncWriteExt as _};
 use tracing::{debug, error, info};
 
+const SUPPORTED_TESTS: &[&str] = &[
```
Comment (djc) [inline quinn-interop/src/client.rs]: Nit: consts near the bottom?

File: quinn-interop/src/server.rs [hunk 6]
```diff
@@ -69,7 +69,11 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let crypto = quinn::crypto::rustls::QuicServerConfig::try_from(crypto)?;
     let mut transport_config = quinn::TransportConfig::default();
     // https://github.com/quic-interop/quic-interop-runner/issues/397
-    transport_config.enable_segmentation_offload(false);
+    transport_config
```
Comment (djc) [inline quinn-interop/src/server.rs]: Should we share some `TransportConfig` setup between client and server?

File: quinn-interop/src/client.rs [hunk 7]
```diff
@@ -27,35 +50,14 @@ use tracing::{debug, error, info};
...
-            std::process::exit(127);
-        }
-        _ => (), // Probably a real testcase run
+    let Ok(testcase) = std::env::var("TESTCASE_CLIENT") else {
+        return ();
+    };
+    if SUPPORTED_TESTS.contains(&&*testcase) {
+        std::process::exit(0);
     }
+    error!("Test case not supported: {}", testcase);
```
Comment (djc) [inline quinn-interop/src/client.rs]: Nit: while this is moved/changed, inline the format argument?

File: quinn-interop/src/client.rs [hunk 8]
```diff
@@ -252,14 +253,8 @@ fn config(alpn: &str, suites: CipherSuite) -> Result<ClientConfig, Box<dyn std::
...
 
-async fn connect(
-    uri: &http::Uri,
-    client_config: ClientConfig,
-) -> Result<(Endpoint, Connection), Box<dyn std::error::Error>> {
-    let auth = uri
-        .authority()
-        .ok_or("destination must have a host")?
-        .clone();
+async fn resolve(uri: &http::Uri) -> Result<(SocketAddr, &str), Box<dyn std::error::Error>> {
```
Comment (djc) [inline quinn-interop/src/client.rs]: Nit: move this below all the `connect_*()` functions?

File: quinn-interop/src/client.rs [hunk 9]
```diff
@@ -113,6 +113,26 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                    Ok::<(), Box<dyn std::error::Error>>(())
+                }
+                .await;
+                if let Err(e) = result {
+                    error!("request failed: {}", e);
+                }
+            });
+        }
+        while let Some(_) = set.join_next().await {}
+        return Ok(());
```
Comment (djc) [inline quinn-interop/src/client.rs]: Shouldn't this be able to yield `Err` somehow in order to propagate test case failure?

File: quinn-interop/src/server.rs [hunk 10]
```diff
@@ -91,11 +91,22 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
             incoming.retry().unwrap();
             continue;
         }
-        let mut new_conn = incoming.accept()?;
+        let mut new_conn = match incoming.accept() {
+            Ok(c) => c,
+            Err(e) => {
+                error!("connection attempt failed: {}", e);
```
Comment (djc) [inline quinn-interop/src/server.rs]: Nit: inline format args?

File: quinn-interop/src/client.rs [hunk 11]
```diff
@@ -62,7 +63,7 @@ fn test_case_implemented_or_exit() {
 }
 
 #[tokio::main(flavor = "multi_thread")]
-async fn main() -> Result<(), Box<dyn std::error::Error>> {
+async fn main() -> Result<()> {
```
Comment (djc) [inline quinn-interop/src/client.rs]: I've never loved shadowing `Result` this way, and usually like to keep it explicitly as `anyhow::Result` in return types. What do you think?

File: quinn-interop/src/server.rs [hunk 12]
```diff
@@ -99,28 +99,19 @@ async fn main() -> Result<()> {
...
-
         tokio::spawn(async move {
+            let handshake_data = match new_conn.handshake_data().await {
+                Ok(x) => x.downcast::<HandshakeData>().unwrap(),
+                Err(e) => {
+                    error!("connection attempt failed: {}", e);
+                    return;
+                }
+            };
+            let alpn = String::from_utf8_lossy(&handshake_data.protocol.as_ref().unwrap());
```
Comment (djc) [inline quinn-interop/src/server.rs]: Pre-existing issue, but it would be better to keep ALPN in the `Vec<u8>` domain instead of converting to `String`.

**Review by djc — APPROVED** [review]
Nice work!

File: quinn-interop/src/client.rs (hunk 1 above)
Comment (Ralith) [reply quinn-interop/src/client.rs]: Moved it to initial-only use instead, for better conformance with the documented requirements for the test.

File: quinn-interop/src/client.rs (hunk 11 above)
Comment (Ralith) [reply quinn-interop/src/client.rs]: I don't feel strongly; switched.

**Pull request merged**